File: 0055-jump-game/0055-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        l = 0
        r = 0
        #print(r)
        while r < len(nums)-1:
            m = 0
            flag = False
            for i in range(l, r+1):
                m = max(m, i+nums[i])
                if flag == False and nums[i] != 0:
                    flag = True
                #print(m)
            l = r+1
            r = m
            if flag == False: 
                return False
        

        return True 
        

File: 0055-jump-game/0055-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        int low = 0, far = nums[0], prevFar = far, n = nums.length;

        while (far < n-1){
            prevFar = far;
            for (int i = low; i <= far && i < n; i++){
                far = Math.max(far, nums[i] + i);
            }
            if (far == prevFar) return false;
            low = prevFar+1;
        }
        return true;
    }
}